lib/src/tokens.dart

```diff
@@ -0,0 +1,143 @@
+// Copyright (c) 2012, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'dart:convert';
+import 'package:path/path.dart' as path;
+
+import 'io.dart';
+import 'log.dart' as log;
+import 'system_cache.dart';
+import 'utils.dart';
+
+List<TokenEntry> _tokens;
+
+/// Gets the token for the given uri
+String getToken(SystemCache cache, Uri uri) {
+  if (uri.host == 'pub.dartlang.org') return null;
+  log.fine('Lookup token for ${uri.toString()}');
+  final tokens = _loadTokens(cache);
+
+  final found = tokens.firstWhere(
+      (e) =>
+          checkEndSlashUri(uri).toString().toLowerCase().startsWith(e.server),
+      orElse: () => null);
+  if (found == null) {
+    log.fine('No token found for ${uri.toString()}');
+    return null;
+  }
+  return found.token;
+}
+
+/// Adds a token for a given server
+void addToken(SystemCache cache, String server, String token) {
+  final tokens = _loadTokens(cache);
+  final normalizedServer = checkEndSlash(server).toLowerCase();
+
+  final found = tokens.firstWhere((e) => e.server == normalizedServer,
+      orElse: () => null);
+  if (found != null) {
+    tokens.remove(found);
+    tokens.add(TokenEntry(server: normalizedServer, token: token));
+    log.message('Token for $server updated');
+  } else {
+    tokens.add(TokenEntry(server: normalizedServer, token: token));
+    log.message('Token for $server added');
+  }
+  _save(cache, tokens);
+}
+
+/// Removes the token for the given server
+void removeToken(SystemCache cache, {String server, bool all = false}) {
+  final tokens = _loadTokens(cache);
+  if (all) {
+    if (tokens.isEmpty) return;
+    for (var item in tokens) {
+      log.message('Log out ${item.server} successful.');
+    }
+    final tokensFile = _tokensFile(cache);
+    if (entryExists(tokensFile)) deleteEntry(tokensFile);
+    return;
+  }
+  final found = tokens.firstWhere(
+      (e) => e.server == checkEndSlash(server).toLowerCase(),
+      orElse: () => null);
+  if (found == null) {
+    log.message('No token found for $server.');
+  } else {
+    tokens.remove(found);
+    log.message('Log out $server successful.');
+  }
+  _save(cache, tokens);
+}
+
+String validateServer(String server) {
```

Comment:
Can we add a documentation comment?

This is for argument validation for commands that take a `--server` argument, right?

Maybe we could call it: `validateCustomPubServerUrl` ? or something even better?